File: 67 Add Binary/67 LL.cpp
#include <iostream>
#include <algorithm>
using namespace std;

/*
https://leetcode.com/problems/add-binary/

Given two binary strings, return their sum (also a binary string).

For example,
a = "11"
b = "1"
Return "100".
*/

class Solution
{
public:
	string addBinary(string a, string b)
	{
		string result;
		string::reverse_iterator va = a.rbegin();
		string::reverse_iterator vb = b.rbegin();
		int carry = 0;
		while (true)
		{
			bool looping = (carry != 0);
			if (va != a.rend())
			{
				looping = true;
				carry += (va[0] == '1') ? 1 : 0;
				++va;
			}
			if (vb != b.rend())
			{
				looping = true;
				carry += (vb[0] == '1') ? 1 : 0;
				++vb;
			}
			if (!looping) break;
			result += '0' + (carry & 0x1);
			carry >>= 1;
		}
		if (result.length() != 0)
		{
			reverse(result.begin(), result.end());
		}
		else result += '0';	// handle empty input
		return result;
	}
};

/*
idea: simulate bit-based adding
complexity: O(N)
*/

int main(void)
{
	Solution engine;
	cout << engine.addBinary("11", "1") << '\n';
	return 0;
}

File: 67 Add Binary/67 LL.py
class Solution(object):
	def addBinary(self, a, b):
		va = [1 if a[i] == '1' else 0 for i in xrange(len(a)-1,-1,-1)]
		vb = [1 if b[i] == '1' else 0 for i in xrange(len(b)-1,-1,-1)]
		pos = 0
		carry = 0
		result = list()
		while True:
			loop = False
			if pos < len(va):
				carry += va[pos]
				loop = True
			if pos < len(vb):
				carry += vb[pos]
				loop = True
			if (not loop) and carry == 0:
				break
			result.append(chr(ord('0') + (carry & 1)))
			carry >>= 1
			pos += 1
		return ''.join(result[::-1])

engine = Solution()
print engine.addBinary("011111", "1")